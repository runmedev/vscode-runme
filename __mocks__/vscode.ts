import path from 'node:path'
import { readFileSync } from 'node:fs'
import { join } from 'node:path'

import { vi } from 'vitest'
import { URI } from 'vscode-uri'
import type { Command, Disposable } from 'vscode'


export const notebooks = {
  createNotebookController: vi.fn().mockReturnValue({
    dispose: vi.fn(),
    createNotebookCellExecution: vi.fn().mockReturnValue({ start: vi.fn(), end: vi.fn() })
  }),
  registerNotebookCellStatusBarItemProvider: vi.fn(),
  createRendererMessaging: vi.fn().mockReturnValue({
    postMessage: vi.fn(),
    onDidReceiveMessage: vi.fn().mockReturnValue({ dispose: vi.fn() })
  }),
}

export class Uri extends URI {
  static file = vi.fn(super.file)
  static parse = vi.fn(super.parse)

  static joinPath = vi.fn((uri: Uri, ...paths: string[]) => {
    return super.file(path.join(uri.fsPath, ...paths))
  })
}

export const workspace = {
  getConfiguration: vi.fn().mockReturnValue(new Map()),
  onDidChangeConfiguration: vi.fn(),
  openNotebookDocument: vi.fn().mockReturnValue({ uri: 'new notebook uri' }),
  openTextDocument: vi.fn().mockReturnValue({
    getText: vi.fn().mockReturnValue(readFileSync(join(__dirname, 'gitignore.mock'), 'utf8'))
  }),
  registerNotebookSerializer: vi.fn(),
  onDidOpenNotebookDocument: vi.fn().mockReturnValue({ dispose: vi.fn() }),
  onDidSaveNotebookDocument: vi.fn().mockReturnValue({ dispose: vi.fn() }),
  onDidChangeNotebookDocument: vi.fn().mockReturnValue({ dispose: vi.fn() }),
  fs: {
    writeFile: vi.fn(),
    readFile: vi.fn().mockResolvedValue(Buffer.from('some wasm file')),
    stat: vi.fn().mockResolvedValue(1),
    createDirectory: vi.fn(),
    delete: vi.fn()
  },
  createFileSystemWatcher: vi.fn().mockReturnValue({
    onDidCreate: vi.fn(),
    onDidDelete: vi.fn()
  }),
  workspaceFolders: [ { uri: { fsPath: '/runme/workspace' } } ],
  findFiles: vi.fn().mockReturnValue([
    {
      path: 'runme/workspace/README.md',
    },
    {
      path: 'runme/workspace/src/README.md',
    },
    {
      path: 'runme/workspace/src/COMMANDS.md',
    },
    {
      path: 'runme/workspace/src/RUNME.md',
    }
  ]),
}

export const terminal = {
  show: vi.fn(),
  sendText: vi.fn(),
  dispose: vi.fn()
}

export const window = {
  createOutputChannel: vi.fn().mockReturnValue({
    appendLine: vi.fn()
  }),
  showWarningMessage: vi.fn(),
  showInformationMessage: vi.fn(),
  showErrorMessage: vi.fn(),
  createTerminal: vi.fn().mockReturnValue(terminal),
  showNotebookDocument: vi.fn(),
  showTextDocument: vi.fn(),
  onDidChangeActiveNotebookEditor: vi.fn().mockReturnValue({ dispose: vi.fn() }),
  registerTreeDataProvider: vi.fn(),
  registerUriHandler: vi.fn(),
  onDidCloseTerminal: vi.fn(),
  withProgress: vi.fn(),
  onDidChangeActiveColorTheme: vi.fn().mockReturnValue({ dispose: vi.fn() }),
}

export const tasks = {
  registerTaskProvider: vi.fn(),
  onDidEndTaskProcess: vi.fn(),
  onDidStartTaskProcess: vi.fn(),
  executeTask: vi.fn().mockResolvedValue(undefined)
}

export const commands = {
  registerCommand: vi.fn(),
  executeCommand: vi.fn(),
  getCommands: vi.fn().mockResolvedValue(['_notebook.selectKernel'])
}

export enum ViewColumn {
  Beside = 'Beside'
}

export const env = {
  clipboard: {
    writeText: vi.fn()
  },
  openExternal: vi.fn()
}

export const NotebookData = vi.fn()
export const NotebookCellData = vi.fn()
export enum NotebookCellKind {
  Code = 1,
  Markup = 2
}

export const TreeItem = vi.fn()
export const Event = vi.fn()

export enum TreeItemCollapsibleState {
  None = 0,
  Collapsed = 1,
  Expanded = 2
}

export enum FileType {
  /**
   * The file type is unknown.
   */
  Unknown = 0,
  /**
   * A regular file.
   */
  File = 1,
  /**
   * A directory.
   */
  Directory = 2,
  /**
   * A symbolic link to a file.
   */
  SymbolicLink = 64
}

export enum NotebookCellStatusBarAlignment {
  Left = 1,
  Right = 2
}

export class NotebookCellStatusBarItem {
  label: string

  alignment: NotebookCellStatusBarAlignment

  tooltip?: string

  priority?: number

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  constructor(label: string, alignment: NotebookCellStatusBarAlignment) {
    this.label = label
    this.alignment = alignment
  }
}



export class NotebookCellOutputItem {

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  static json(value: any, mime = 'text/plain'): NotebookCellOutputItem {
    return {
      mime: mime,
      data: new Uint8Array()
    }
  }

  mime: string
  data: Uint8Array
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  constructor(data: Uint8Array, mime: string) {}
}

export class NotebookCellOutput {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  constructor(public items: NotebookCellOutputItem[], public metadata?: { [key: string]: any }) {}
}

export const ProgressLocation = {
  Window: 1
}

type MessageCallback<T> = (message: T) => void
type Event<T> = (listener: MessageCallback<T>) => Disposable

export class EventEmitter<T> {
  listeners: MessageCallback<T>[] = []

  event: Event<T> = (listener) => {
    this.listeners.push(listener)

    return {
      dispose: () => {
        this.listeners = this.listeners.filter(x => x !== listener)
      }
    }
  }

  fire(data: T) {
    this.listeners.forEach(l => l(data))
  }

  async fireAsync(data: T) {
    await Promise.all(
      this.listeners.map(l => l(data))
    )
  }

  dispose() { }
}

export const languages = {
  registerCodeLensProvider: vi.fn(),
}

export class Position {
  constructor(
    public line: number,
    public character: number
  ) { }

  with(line?: number, character?: number): Position {
    return new Position(
      line ?? this.line,
      character ?? this.character,
    )
  }
}

export class Range {
  constructor(
    public start: Position,
    public end: Position,
  ) { }
}

export class CodeLens {
  constructor(
    public range: Range,
    public command?: Command
  ) { }
}

export enum TaskScope {
  /**
   * The task is a global task. Global tasks are currently not supported.
   */
  Global = 1,

  /**
   * The task is a workspace task
   */
  Workspace = 2
}

export const ShellExecution = vi.fn()
export const Task = vi.fn()
