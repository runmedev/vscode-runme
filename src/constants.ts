/**
 * Note: this file is used within Node.js and Browser environment.
 * Only export cross compatible objects here.
 */

export enum OutputType {
  vercel = 'stateful.runme/vercel-stdout',
  deno = 'stateful.runme/deno-stdout',
  outputItems = 'stateful.runme/output-items',
  annotations = 'stateful.runme/annotations',
  error = 'stateful.runme/error'
}

export enum ClientMessages {
  infoMessage = 'common:infoMessage',
  errorMessage = 'common:errorMessage',
  update = 'deno:deploymentUpdate',
  deployed = 'deno:finishedDeployment',
  promote = 'deno:promoteDeployment',
  prod = 'vercel:promotePreview',
  mutateAnnotations = 'annotations:mutate'
}

// [pretty print, languageId, destination]
export const LANGUAGES = new Map([
  ['Assembly', 'asm', 'sh'],
  ['Batchfile', 'bat', 'sh'],
  ['C', 'c', undefined],
  ['C#', 'cs', undefined],
  ['C++', 'cpp', undefined],
  ['Clojure', 'clj', undefined],
  ['CMake', 'cmake', 'sh'],
  ['COBOL', 'cbl', 'null'],
  ['CoffeeScript', 'coffee', 'html'],
  ['CSS', 'css', undefined],
  ['CSV', 'csv', undefined],
  ['Dart', 'dart', undefined],
  ['DM', 'dm', undefined],
  ['Dockerfile', 'dockerfile', 'sh'],
  ['Elixir', 'ex', undefined],
  ['Erlang', 'erl', undefined],
  ['Fortran', 'f90', undefined],
  ['Go', 'go', undefined],
  ['Groovy', 'groovy', undefined],
  ['Haskell', 'hs', undefined],
  ['HTML', 'html', undefined],
  ['INI', 'ini', undefined],
  ['Java', 'java', undefined],
  ['JavaScript', 'js', undefined],
  ['JSON', 'json', undefined],
  ['Julia', 'jl', undefined],
  ['Kotlin', 'kt', undefined],
  ['Lisp', 'lisp', undefined],
  ['Lua', 'lua', undefined],
  ['Makefile', 'makefile', 'sh'],
  ['Markdown', 'md', undefined],
  ['Matlab', 'matlab', undefined],
  ['Objective-C', 'mm', undefined],
  ['OCaml', 'ml', undefined],
  ['Pascal', 'pas', undefined],
  ['Perl', 'pm', undefined],
  ['PHP', 'php', undefined],
  ['PowerShell', 'ps1', undefined],
  ['Prolog', 'prolog', undefined],
  ['Python', 'py', undefined],
  ['R', 'r', undefined],
  ['Ruby', 'rb', undefined],
  ['Rust', 'rs', undefined],
  ['Scala', 'scala', undefined],
  ['Shell', 'sh', 'sh'],
  ['SQL', 'sql', undefined],
  ['Swift', 'swift', undefined],
  ['TeX', 'tex', undefined],
  ['TOML', 'toml', undefined],
  ['TypeScript', 'ts', undefined],
  ['Verilog', 'v', undefined],
  ['Visual Basic', 'vba', undefined],
  ['XML', 'xml', undefined],
  ['YAML', 'yaml', undefined],
].map(([, source, dest]) => [source, dest]))
