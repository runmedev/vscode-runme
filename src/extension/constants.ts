import os from 'node:os'

export const PLATFORM_OS = os.platform()
export const DENO_ACCESS_TOKEN_KEY = 'DENO_ACCESS_TOKEN'
export const DENO_PROJECT_NAME_KEY = 'DENO_PROJECT_NAME'

export const DEFAULT_ENV = {
  RUNME_TASK: 'true',
  PATH: process.env.PATH || ''
}
export const ENV_STORE = new Map<string, string>(
  Object.entries(DEFAULT_ENV)
)

export const BOOTFILE = '.runme_bootstrap'
