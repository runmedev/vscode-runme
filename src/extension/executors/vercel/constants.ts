export const VERCEL_DIR = '.vercel'
