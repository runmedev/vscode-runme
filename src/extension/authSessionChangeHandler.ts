import { Subject, Subscription } from 'rxjs'
import { debounceTime, distinctUntilChanged } from 'rxjs/operators'
import { authentication, AuthenticationSessionsChangeEvent, Disposable } from 'vscode'

export default class AuthSessionChangeHandler implements Disposable {
  static #instance: AuthSessionChangeHandler | null = null

  #disposables: Disposable[] = []
  #eventSubject: Subject<AuthenticationSessionsChangeEvent>
  #subscriptions: Subscription[] = []
  #listeners: ((event: AuthenticationSessionsChangeEvent) => void)[] = []

  private constructor(private debounceTimeMs: number = 100) {
    this.#eventSubject = new Subject<AuthenticationSessionsChangeEvent>()
    this.#subscriptions.push(
      this.#eventSubject
        .pipe(distinctUntilChanged(this.eventComparer), debounceTime(this.debounceTimeMs))
        .subscribe((event) => {
          this.notifyListeners(event)
        }),
    )

    this.#disposables.push(
      authentication.onDidChangeSessions((e) => {
        this.#eventSubject.next(e)
      }),
    )
  }

  public static get instance(): AuthSessionChangeHandler {
    if (!this.#instance) {
      this.#instance = new AuthSessionChangeHandler()
    }

    return this.#instance
  }

  public addListener(listener: (event: AuthenticationSessionsChangeEvent) => void): void {
    this.#listeners.push(listener)
  }

  public removeListener(listener: (event: AuthenticationSessionsChangeEvent) => void): void {
    this.#listeners = this.#listeners.filter((l) => l !== listener)
  }

  private notifyListeners(event: AuthenticationSessionsChangeEvent): void {
    for (const listener of this.#listeners) {
      try {
        listener(event)
      } catch (err) {
        console.error('Error in listener:', err)
      }
    }
  }

  private eventComparer(
    previous: AuthenticationSessionsChangeEvent,
    current: AuthenticationSessionsChangeEvent,
  ): boolean {
    return (
      previous.provider.id === current.provider.id &&
      JSON.stringify(previous) === JSON.stringify(current)
    )
  }

  public async dispose() {
    this.#disposables.forEach((d) => d.dispose())
    this.#subscriptions = []
    this.#eventSubject.complete()
    this.#listeners = []

    AuthSessionChangeHandler.#instance = null
  }
}
