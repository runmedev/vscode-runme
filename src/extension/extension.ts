import { workspace, notebooks, commands, ExtensionContext, tasks, window, Uri } from 'vscode'
import { TelemetryReporter } from 'vscode-telemetry'

import { Serializer } from '../types'
import { registerExtensionEnvironmentVariables } from '../utils/configuration'

import { Kernel } from './kernel'
import RunmeServer from './server/runmeServer'
import RunmeServerError from './server/runmeServerError'
import { ToggleTerminalProvider, BackgroundTaskProvider, StopBackgroundTaskProvider } from './provider/background'
import { CopyProvider } from './provider/copy'
import { getDefaultWorkspace, resetEnv } from './utils'
import { AnnotationsProvider } from './provider/annotations'
import { RunmeTaskProvider } from './provider/runmeTask'
import {
  toggleTerminal,
  runCLICommand,
  copyCellToClipboard,
  openAsRunmeNotebook,
  openSplitViewAsMarkdownText,
  stopBackgroundTask,
  createNewRunmeNotebook,
  welcome,
  tryIt,
  openFileInRunme,
  openIntegratedTerminal,
  authenticateWithGitHub,
  displayCategoriesSelector,
  runCellsByCategory
} from './commands'
import { WasmSerializer, GrpcSerializer } from './serializer'
import { RunmeLauncherProvider } from './provider/launcher'
import { RunmeUriHandler } from './handler/uri'
import { BOOTFILE } from './constants'
import { GrpcRunner, IRunner } from './runner'
import { CliProvider } from './provider/cli'
import * as survey from './survey'
import { RunmeCodeLensProvider } from './provider/codelens'

export class RunmeExtension {
  async initialize(context: ExtensionContext) {
    const kernel = new Kernel(context)
    const grpcSerializer = kernel.hasExperimentEnabled('grpcSerializer')
    const grpcServer = kernel.hasExperimentEnabled('grpcServer')
    const grpcRunner = kernel.hasExperimentEnabled('grpcRunner')
    const server = new RunmeServer(context.extensionUri, {
      retryOnFailure: true,
      maxNumberOfIntents: 10,
    }, !grpcServer, grpcRunner)

    let runner: IRunner | undefined
    if (grpcRunner) {
      runner = new GrpcRunner(server)
      kernel.useRunner(runner)
    }

    const serializer = grpcSerializer ?
      new GrpcSerializer(context, server, kernel) :
      new WasmSerializer(context, kernel)

    /**
     * Start the Runme server
     */
    try {
      await server.launch()
    } catch (e) {
      // Unrecoverable error happened
      if (e instanceof RunmeServerError) {
        TelemetryReporter.sendTelemetryErrorEvent('extension.server', { data: e.message })
        return window.showErrorMessage(`Failed to start Runme server. Reason: ${e.message}`)
      }
      TelemetryReporter.sendTelemetryErrorEvent('extension.server', { data: (e as Error).message })
      return window.showErrorMessage(
        'Failed to start Runme server, please try to reload the window. ' +
        `Reason: ${(e as any).message}`
      )
    }

    const treeViewer = new RunmeLauncherProvider(getDefaultWorkspace())
    const uriHandler = new RunmeUriHandler(context)
    const activeUsersSurvey = new survey.SurveyActiveUserFeedback(context)
    const winCodeLensRunSurvey = new survey.SurveyWinCodeLensRun(context)
    const stopBackgroundTaskProvider = new StopBackgroundTaskProvider()

    const runCLI = runCLICommand(context.extensionUri, !!grpcRunner, server, kernel)

    const codeLensProvider = new RunmeCodeLensProvider(serializer, runCLI, winCodeLensRunSurvey, runner, kernel)

    registerExtensionEnvironmentVariables(context)

    const omitKeys: Serializer.Metadata = {
      ['runme.dev/name']: undefined,
      ['runme.dev/uuid']: undefined,
      ['runme.dev/textRange']: undefined,
    }
    const transientCellMetadata = Object.fromEntries(
      Object.keys(omitKeys).map((k) => [k, true])
    )

    context.subscriptions.push(
      kernel,
      serializer,
      server,
      treeViewer,
      activeUsersSurvey,
      winCodeLensRunSurvey,
      workspace.registerNotebookSerializer(Kernel.type, serializer, {
        transientOutputs: true,
        transientCellMetadata
      }),

      notebooks.registerNotebookCellStatusBarItemProvider(Kernel.type, new ToggleTerminalProvider(kernel)),
      notebooks.registerNotebookCellStatusBarItemProvider(Kernel.type, new BackgroundTaskProvider()),
      notebooks.registerNotebookCellStatusBarItemProvider(Kernel.type, new CopyProvider()),
      notebooks.registerNotebookCellStatusBarItemProvider(Kernel.type, stopBackgroundTaskProvider),
      notebooks.registerNotebookCellStatusBarItemProvider(Kernel.type, new AnnotationsProvider(kernel)),

      stopBackgroundTaskProvider,

      codeLensProvider,

      commands.registerCommand('runme.resetEnv', resetEnv),
      RunmeExtension.registerCommand('runme.openIntegratedTerminal', openIntegratedTerminal),
      RunmeExtension.registerCommand('runme.toggleTerminal', toggleTerminal(kernel, !!grpcRunner)),
      RunmeExtension.registerCommand(
        'runme.runCliCommand',
        runCLI
      ),
      RunmeExtension.registerCommand('runme.copyCellToClipboard', copyCellToClipboard),
      RunmeExtension.registerCommand('runme.stopBackgroundTask', stopBackgroundTask),
      RunmeExtension.registerCommand('runme.openSplitViewAsMarkdownText', openSplitViewAsMarkdownText),
      RunmeExtension.registerCommand('runme.openAsRunmeNotebook', openAsRunmeNotebook),
      RunmeExtension.registerCommand('runme.runCategory', (notebook) => {
        return displayCategoriesSelector({ context, kernel, notebookToolbarCommand: notebook })
      }),
      RunmeExtension.registerCommand('runme.runCellCategory', (cell) => runCellsByCategory(cell, kernel)),
      RunmeExtension.registerCommand('runme.new', createNewRunmeNotebook),
      RunmeExtension.registerCommand('runme.welcome', welcome),
      RunmeExtension.registerCommand('runme.try', () => tryIt(context)),
      RunmeExtension.registerCommand('runme.openRunmeFile', RunmeLauncherProvider.openFile),
      RunmeExtension.registerCommand('runme.keybinding.m', () => { }),
      RunmeExtension.registerCommand('runme.keybinding.y', () => { }),
      RunmeExtension.registerCommand('runme.file.openInRunme', openFileInRunme),
      tasks.registerTaskProvider(RunmeTaskProvider.id, new RunmeTaskProvider(context, serializer, runner, kernel)),
      notebooks.registerNotebookCellStatusBarItemProvider(Kernel.type, new CliProvider()),

      /**
       * tree viewer items
       */
      window.registerTreeDataProvider('runme.launcher', treeViewer),
      RunmeExtension.registerCommand('runme.collapseTreeView', treeViewer.collapseAll.bind(treeViewer)),
      RunmeExtension.registerCommand('runme.expandTreeView', treeViewer.expandAll.bind(treeViewer)),
      RunmeExtension.registerCommand('runme.authenticateWithGitHub', authenticateWithGitHub),
      /**
       * Uri handler
       */
      window.registerUriHandler(uriHandler)
    )

    if (workspace.workspaceFolders?.length && workspace.workspaceFolders[0]) {
      const startupFileUri = Uri.joinPath(workspace.workspaceFolders[0].uri, BOOTFILE)
      const hasStartupFile = await workspace.fs.stat(startupFileUri).then(() => true, () => false)
      if (hasStartupFile) {
        const bootFile = new TextDecoder().decode(await workspace.fs.readFile(startupFileUri))
        const bootFileUri = Uri.joinPath(workspace.workspaceFolders[0].uri, bootFile)
        await workspace.fs.delete(startupFileUri)
        await commands.executeCommand('vscode.openWith', bootFileUri, Kernel.type)
      }
    }
  }

  static registerCommand(command: string, callback: (...args: any[]) => any, thisArg?: any) {
    return commands.registerCommand(command, (...wrappedArgs: any[]) => {
      TelemetryReporter.sendTelemetryEvent('extension.command', { command })
      return callback(...wrappedArgs, thisArg)
    }, thisArg)
  }
}
