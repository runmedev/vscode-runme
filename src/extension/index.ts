import type { ExtensionContext } from 'vscode'
import { TelemetryReporter } from 'vscode-telemetry'

import { RunmeExtension } from './extension'
import getLogger from './logger'

declare const INSTRUMENTATION_KEY: string

const ext = new RunmeExtension()
const log = getLogger()

export async function activate (context: ExtensionContext) {
  TelemetryReporter.configure(context, INSTRUMENTATION_KEY)
  log.info('Activating Extension')
  try {
    await ext.initialize(context)
    log.info('Extension successfully activated')
  } catch (err: any) {
    log.error(`Failed to initialize the extension: ${err.message}`)
  }

  TelemetryReporter.sendTelemetryEvent('activate')
}

export function deactivate () {
  log.info('Deactivating Extension')
  TelemetryReporter.sendTelemetryEvent('deactivate')
}
