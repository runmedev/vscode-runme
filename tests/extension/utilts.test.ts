import vscode from 'vscode'
import { expect, vi, test, beforeAll, afterAll, suite } from 'vitest'

import {
  getExecutionProperty,
  getTerminalByCell,
  resetEnv,
  getKey,
  getCmdShellSeq,
  normalizeLanguage,
} from '../../src/extension/utils'
import { ENV_STORE, DEFAULT_ENV } from '../../src/extension/constants'

vi.mock('vscode', () => ({
  default: {
    window: {
      terminals: [
        { creationOptions: { env: {} } },
        { creationOptions: { env: { RUNME_ID: 'foobar:123' } } }
      ]
    },
    workspace: {
      getConfiguration: vi.fn()
    }
  }
}))

const PATH = process.env.PATH
beforeAll(() => {
  DEFAULT_ENV.PATH = '/usr/bin'
  ENV_STORE.delete('PATH')
})
afterAll(() => { process.env.PATH = PATH })

test('isInteractive', () => {
  // when set to false in configutaration
  vi.mocked(vscode.workspace.getConfiguration).mockReturnValue({ get: vi.fn().mockReturnValue(false) } as any)
  expect(getExecutionProperty('interactive', { metadata: {} } as any)).toBe(false)
  expect(getExecutionProperty('interactive', { metadata: { attributes: {} } } as any)).toBe(false)
  expect(getExecutionProperty('interactive', { metadata: { attributes: { interactive: 'true' } } } as any)).toBe(true)

  vi.mocked(vscode.workspace.getConfiguration).mockReturnValue({ get: vi.fn().mockReturnValue(true) } as any)
  expect(getExecutionProperty('interactive', { metadata: {} } as any)).toBe(true)
  expect(getExecutionProperty('interactive', { metadata: { attributes: {} } } as any)).toBe(true)
})

test('getTerminalByCell', () => {
  expect(getTerminalByCell({ document: { fileName: 'foo' }, index: 42} as any))
    .toBe(undefined)
  expect(getTerminalByCell({ document: { fileName: 'foobar' }, index: 123} as any))
    .not.toBe(undefined)
})

test('resetEnv', () => {
  ENV_STORE.set('foo', 'bar')
  expect(ENV_STORE).toMatchSnapshot()
  resetEnv()
  expect(ENV_STORE).toMatchSnapshot()
})

test('getKey', () => {
  expect(getKey({
    getText: vi.fn().mockReturnValue('foobar'),
    languageId: 'barfoo'
  } as any)).toBe('barfoo')
  expect(getKey({
    getText: vi.fn().mockReturnValue('deployctl deploy foobar'),
    languageId: 'something else'
  } as any)).toBe('deno')
})

suite('getCmdShellSeq', () => {
  test('one command', () => {
    const cellText = 'deno task start'
    expect(getCmdShellSeq(cellText, 'darwin')).toMatchSnapshot()
  })

  test('wrapped command', () => {
    // eslint-disable-next-line max-len
    const cellText = Buffer.from('ZGVubyBpbnN0YWxsIFwKICAgICAgLS1hbGxvdy1yZWFkIC0tYWxsb3ctd3JpdGUgXAogICAgICAtLWFsbG93LWVudiAtLWFsbG93LW5ldCAtLWFsbG93LXJ1biBcCiAgICAgIC0tbm8tY2hlY2sgXAogICAgICAtciAtZiBodHRwczovL2Rlbm8ubGFuZC94L2RlcGxveS9kZXBsb3ljdGwudHMK', 'base64').toString('utf-8')

    expect(getCmdShellSeq(cellText, 'darwin')).toMatchSnapshot()
  })

  test('env only', () => {
    const cellText = `export DENO_INSTALL="$HOME/.deno"
      export PATH="$DENO_INSTALL/bin:$PATH"
    `
    expect(getCmdShellSeq(cellText, 'darwin')).toMatchSnapshot()
  })

  test('complex wrapped', () => {
    // eslint-disable-next-line max-len
    const cellText = 'curl "https://api-us-west-2.graphcms.com/v2/cksds5im94b3w01xq4hfka1r4/master?query=$(deno run -A query.ts)" --compressed 2>/dev/null \\\n| jq -r \'.[].posts[] | "\(.title) - by \(.authors[0].name), id: \(.id)"\''
    expect(getCmdShellSeq(cellText, 'darwin')).toMatchSnapshot()
  })

  test('linux without pipefail', () => {
    const cellText = 'ls ~/'
    expect(getCmdShellSeq(cellText, 'linux')).toMatchSnapshot()
  })

  test('windows without shell flags', () => {
    const cellText = 'ls ~/'
    expect(getCmdShellSeq(cellText, 'win32')).toMatchSnapshot()
  })

  test('with comments', () => {
    // eslint-disable-next-line max-len
    const cellText = 'echo "Install deno via installer script"\n# macOS or Linux\ncurl -fsSL https://deno.land/x/install/install.sh | sh'
    expect(getCmdShellSeq(cellText, 'darwin')).toMatchSnapshot()
  })

  test('trailing comment', () => {
    const cellText = 'cd ..\nls / # list dir contents\ncd ..\nls /'
    expect(getCmdShellSeq(cellText, 'darwin')).toMatchSnapshot()
  })
})

suite('normalizeLanguage', () => {
  test('with zsh', () => {
    const lang = normalizeLanguage('zsh')
    expect(lang).toBe('sh')
  })

  test('with shell', () => {
    const lang = normalizeLanguage('shell')
    expect(lang).toBe('sh')
  })

  test('with sh', () => {
    const lang = normalizeLanguage('sh')
    expect(lang).toBe('sh')
  })
})
