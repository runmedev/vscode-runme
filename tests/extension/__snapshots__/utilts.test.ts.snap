// Vitest Snapshot v1

exports[`getCmdShellSeq > complex wrapped 1`] = `"set -e -o pipefail; curl \\"https://api-us-west-2.graphcms.com/v2/cksds5im94b3w01xq4hfka1r4/master?query=$(deno run -A query.ts)\\" --compressed 2>/dev/null | jq -r '.[].posts[] | \\"(.title) - by (.authors[0].name), id: (.id)\\"'"`;

exports[`getCmdShellSeq > env only 1`] = `"set -e -o pipefail; export DENO_INSTALL=\\"$HOME/.deno\\"; export PATH=\\"$DENO_INSTALL/bin:$PATH\\""`;

exports[`getCmdShellSeq > linux without pipefail 1`] = `"set -e; ls ~/"`;

exports[`getCmdShellSeq > one command 1`] = `"set -e -o pipefail; deno task start"`;

exports[`getCmdShellSeq > with comments 1`] = `"set -e -o pipefail; echo \\"Install deno via installer script\\"; curl -fsSL https://deno.land/x/install/install.sh | sh"`;

exports[`getCmdShellSeq > wrapped command 1`] = `"set -e -o pipefail; deno install --allow-read --allow-write --allow-env --allow-net --allow-run --no-check -r -f https://deno.land/x/deploy/deployctl.ts"`;

exports[`resetEnv 1`] = `
Map {
  "RUNME_TASK" => "true",
  "foo" => "bar",
}
`;

exports[`resetEnv 2`] = `
Map {
  "RUNME_TASK" => "true",
  "PATH" => "/usr/bin",
}
`;
